File: Python/1044-shopping_in_mars.py
from collections import namedtuple


Solution = namedtuple('Solution', ['bgn', 'end'])


def shopping_in_mars(diamonds, amount):
    solutions = []
    bgn = end = 0
    p_sum = 0
    nearest_amount = float('inf')
    while end < len(diamonds):
        while end < len(diamonds) and p_sum < amount:
            p_sum += diamonds[end]
            end += 1
        while p_sum >= amount:
            if p_sum == amount:
                solutions.append(Solution(bgn, end - 1))
            else:
                if p_sum < nearest_amount:
                    nearest_amount = p_sum
                    nearest_solutions = [Solution(bgn, end - 1)]
                elif p_sum == nearest_amount:
                    nearest_solutions.append(Solution(bgn, end - 1))
            p_sum -= diamonds[bgn]
            bgn += 1
    if solutions:
        return solutions
    else:
        return nearest_solutions


if __name__ == '__main__':
    _, amount = [int(i) for i in input().split()]
    diamonds = [int(i) for i in input().split()]
    solutions = shopping_in_mars(diamonds, amount)
    for bgn, end in solutions:
        print(f'{bgn + 1}-{end + 1}')

File: C++/1044-shopping_in_mars.cc
#include <iostream>
#include <vector>
#include <limits>


struct Solution {
    std::size_t bgn, end;
};


std::vector<Solution>
shopping_in_mars(std::vector<int> &diamonds, int amount)
{
    std::vector<Solution> solutions;
    std::vector<Solution> nearest_solutions;
    int nearest_amount = std::numeric_limits<int>::max();

    std::size_t bgn = 0, end = 0;
    int part_sum = 0;
    while (end < diamonds.size()) {
        while (end < diamonds.size() and part_sum < amount) {
            part_sum += diamonds[end++];
        }
        while (part_sum >= amount) {
            if (part_sum == amount) {
                solutions.push_back({bgn, end - 1});
            } else {
                if (part_sum < nearest_amount) {
                    nearest_amount = part_sum;
                    nearest_solutions.clear();
                    nearest_solutions.push_back({bgn, end - 1});
                } else if (part_sum == nearest_amount) {
                    nearest_solutions.push_back({bgn, end - 1});
                }
            }
            part_sum -= diamonds[bgn++];
        }
    }
    if (!solutions.empty()) {
        return solutions;
    } else {
        return nearest_solutions;
    }
}


int main() {

    std::ios::sync_with_stdio(false);

    int n_diamond, amount;
    std::cin >> n_diamond >> amount;

    std::vector<int> diamonds;
    while (n_diamond--) {
        int diamond;
        std::cin >> diamond;
        diamonds.push_back(diamond);
    }

    std::vector<Solution> solutions = shopping_in_mars(diamonds, amount);
    for (auto &solution : solutions) {
        std::cout << solution.bgn + 1 << '-'
                  << solution.end + 1 << '\n';
    }

    return 0;
}
